### ISSUE TITLE
False positive for redundant field names when using range?

### ISSUE THREAD

**Issue by avranju** [rust-lang-nursery/rust-clippy#2799]
For the following [code](https://play.rust-lang.org/?gist=8ce4e13a9494c3260c7ae3b9a8ec4534&version=stable&mode=debug):
```rust
fn main() {
    let start = 0;
    let fin = 10;
    let vals: Vec<i32> = (start..fin).map(|v| v * v).collect();
    println!("{:?}", vals);
}
```
Clippy warns:
```
warning: redundant field names in struct initialization
 --> src/main.rs:4:27
  |
4 |     let vals: Vec<i32> = (start..fin).map(|v| v * v).collect();
  |                           ^^^^^ help: replace it with: `start`
  |
  = note: #[warn(redundant_field_names)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.204/index.html#redundant_field_names
```

Is this a legitimate lint?

**Comment by brinchj** [comment]
Hi!

As far as I can tell, it looks like this pull request tried to address the issue:
https://github.com/rust-lang-nursery/rust-clippy/pull/2507

But I can still reproduce it in the Rust playground (tools -> clippy):
https://play.rust-lang.org/?gist=8e1d9d87643e9799fce1c60303ea3bab

**Comment by brinchj** [comment]
I can reproduce this locally with Clippy 0.0.198.

**Comment by ordovicia** [comment]
Latest clippy (v0.0.204) does *not* lint `start..end`,
but does `(start..end).map(|v| v * v).collect::<Vec<i32>>()`.

https://play.rust-lang.org/?gist=3cbade6f25fd49565712e71d0a32de78&version=stable&mode=debug

**Comment by mikerite** [comment]
I think this is a compiler bug.

When lowering expressions in the compiler, it looks like they go to a lot of trouble to set up a span with the correct context and then they replace it if it's enclosed in a parenthesis. See [this branch](https://github.com/rust-lang/rust/blob/edae1cc38b467518a8eea590c9c3e0c103b4ecb0/src/librustc/hir/lowering.rs#L3142) and [this branch](https://github.com/rust-lang/rust/blob/edae1cc38b467518a8eea590c9c3e0c103b4ecb0/src/librustc/hir/lowering.rs#L3264) in `lower_expr`. :thinking:

**Comment by flip1995** [comment]
There should be a test for this, before closing this issue.

**Comment by mikerite** [comment]
Unless I'm mistaken...

There are already tests.
https://github.com/rust-lang-nursery/rust-clippy/blob/b109e1033476390f95c728d674909ef90f0ff0ca/tests/ui/redundant_field_names.rs#L42-L51

The errors were fixed by the commit I referenced.
https://github.com/rust-lang-nursery/rust-clippy/pull/3007/files#diff-09e689a73333ff863d80e74c8d58234f

**Comment by flip1995** [comment]
These tests are for the unparenthesized case. The bug in this issue is that it didn't work with `(start..end)`

**Comment by mikerite** [comment]
OK, I'll submit the PR after the build issues are sorted out.
